### DISCUSSION TITLE
Share code between hooks with common args

### DISCUSSION THREAD

**Discussion by andreaterzolo** [bpftrace/bpftrace#3589] [Ideas]
I'm hooking into `kretfunc:tcp_sendmsg` and `kretfunc:tcp_recvmsg` and my 2 programs share almost the same code (this is just an example):

```
#!/usr/bin/bpftrace

kretfunc:tcp_sendmsg
{
    $inet = (struct inet_sock *)args->sk;
    $lport = bswap($inet->inet_sport);
    $dport = bswap(args->sk->__sk_common.skc_dport);

    printf("%d<-%d\n", $dport, $lport);
}

kretfunc:tcp_recvmsg
{
    $inet = (struct inet_sock *)args->sk;
    $lport = bswap($inet->inet_sport);
    $dport = bswap(args->sk->__sk_common.skc_dport);

    printf("%d->%d\n", $dport, $lport);
}
```

Is there a way to share the code between the functions in a clear way? (without creating a monoline macro). Doing something like this 

```
kretfunc:tcp_sendmsg,
kretfunc:tcp_recvmsg
{
    $inet = (struct inet_sock *)args->sk;
    $lport = bswap($inet->inet_sport);
    $dport = bswap(args->sk->__sk_common.skc_dport);

    printf("%d<-%d\n", $dport, $lport);
}
```

returns an error `ERROR: Probe has attach points with mixed arguments` and this is correct, but in the end, i'm just using an argument that both attach points have... Any idea on how to do that?

**Comment by andreaterzolo** [comment]
Seems related to https://github.com/bpftrace/bpftrace/issues/2516 but not sure if that design will address scenarios in which arguments are not constant at compile time

**Reply by jordalgo** [reply]
@Andreagit97 Yes, I believe the PR you linked to (user-defined functions) will solve this and you'll be able to just pass the specific types to a shared function. That being said, since your code is only accessing shared arguments of the same type, I think we should support your use-case without user-defined functions. Do you mind filing an issue - even if we end up using user-defined functions to solve it?

**Reply by andreaterzolo** [reply]
sure! Thank you for the quick feedback!

**Comment by viktormalik** [comment]
The discussion will follow in #3590.
